### DISCUSSION TITLE
"fatal error: 'future' file not found" when make

### DISCUSSION THREAD

**Discussion by Dexter-Wx** [ggml-org/llama.cpp#8201] [Q&A] [answered]
# Got error when build llama.cpp on Macbook m1pro (MacOS15.0 beta2):

## code in terminal:
git clone https://github.com/ggerganov/llama.cpp
cd llama.cpp
make

## the output after make:
dexter@192 /users/dexter/miniforge3/envs/llama.cpp % make
I ccache not found. Consider installing it for faster compilation.
I llama.cpp build info: 
I UNAME_S:   Darwin
I UNAME_P:   arm
I UNAME_M:   arm64
I CFLAGS:    -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_LLAMAFILE -DGGML_USE_METAL -DGGML_METAL_EMBED_LIBRARY  -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion 
I CXXFLAGS:  -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread   -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_LLAMAFILE -DGGML_USE_METAL -DGGML_METAL_EMBED_LIBRARY 
I NVCCFLAGS: -std=c++11 -O3 
I LDFLAGS:   -framework Accelerate -framework Foundation -framework Metal -framework MetalKit 
I CC:        Apple clang version 16.0.0 (clang-1600.0.21.4)
I CXX:       Apple clang version 16.0.0 (clang-1600.0.21.4)

c++ -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread   -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_LLAMAFILE -DGGML_USE_METAL -DGGML_METAL_EMBED_LIBRARY  -c ggml/src/ggml-blas.cpp -o ggml/src/ggml-blas.o
ggml/src/ggml-blas.cpp:4:10: fatal error: 'future' file not found
    4 | #include <future>
      |          ^~~~~~~~
1 error generated.
make: *** [ggml/src/ggml-blas.o] Error 1

## the 2 solutions I have tried but failed
I search my computer and find the file "future" in "/Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1"

1. change #include <future> to #include "future" in ggml/src/ggml-blas.cpp
2. make -I /Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1

**Comment by micsthepick** [comment]
perhaps try the following?

```
mkdir build
cd build
cmake ..
make
```

**Reply by Dexter-Wx** [reply]
thanks but still error.

**Comment by Dexter-Wx** [comment, marked as answer]
Error fixed!
We should save the path of c++ library in environment variable:
export CXXFLAGS=-I/Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1
Or use another way:
Change a variable in the file "Makefile" of llama.cpp:
add -I/Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk/usr/include/c++/v1 behind
MK_CPPFLAGS  = -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon

**Comment by slaren** [comment]
`future` is part of the C++ standard library, there is something wrong with your build environment if your compiler cannot find it.

**Reply by Dexter-Wx** [reply]
yes, that's right.

**Comment by exhyy** [comment]
Same error, I fixed it by re-installing CommandLineTools:
```bash
sudo rm -rf /Library/Developer/CommandLineTools
xcode-select --install
```
It might be because I upgraded CommandLineTools from version 15.0 to 16.0, which caused the error.

**Reply by sebastianczech** [reply]
> Same error, I fixed it by re-installing CommandLineTools:
> 
> ```shell
> sudo rm -rf /Library/Developer/CommandLineTools
> xcode-select --install
> ```
> 
> It might be because I upgraded CommandLineTools from version 15.0 to 16.0, which caused the error.

It helped me too, thank you :)
